src/Workspaces/CoreTestUtilities/ExportProviderCache.cs

```diff
@@ -103,45 +139,67 @@ public static ComposableCatalog WithPart(this ComposableCatalog catalog, Type t)
+            return new SingleExportProviderFactory(catalog, exportProviderFactory);
+        }
+
+        private class SingleExportProviderFactory : IExportProviderFactory
+        {
+            private readonly ComposableCatalog _catalog;
+            private readonly IExportProviderFactory _exportProviderFactory;
+
+            public SingleExportProviderFactory(ComposableCatalog catalog, IExportProviderFactory exportProviderFactory)
             {
-                throw new InvalidOperationException($"Only one {nameof(ExportProvider)} can be created for a single test.");
+                _catalog = catalog;
+                _exportProviderFactory = exportProviderFactory;
             }
 
-            var expected = _expectedProviderForCatalog;
-            if (expected == null)
+            public ExportProvider CreateExportProvider()
             {
-                if (!_exportProviderFactoryCache.TryGetValue(catalog, out var exportProviderFactory))
+                if (!_enabled)
                 {
-                    var configuration = CompositionConfiguration.Create(catalog.WithCompositionService());
-                    var runtimeComposition = RuntimeComposition.CreateRuntimeComposition(configuration);
-                    exportProviderFactory = runtimeComposition.CreateExportProviderFactory();
-                    if (cacheExportProviderFactory)
-                    {
-                        _exportProviderFactoryCache.Add(catalog, exportProviderFactory);
-                    }
+                    throw new InvalidOperationException($"{nameof(ExportProviderCache)} may only be used from tests marked with {nameof(UseExportProviderAttribute)}");
                 }
 
-                expected = exportProviderFactory.CreateExportProvider();
-                expected = Interlocked.CompareExchange(ref _expectedProviderForCatalog, expected, null) ?? expected;
-                Interlocked.CompareExchange(ref _currentExportProvider, expected, null);
-            }
+                var expectedCatalog = Interlocked.CompareExchange(ref _expectedCatalog, _catalog, null) ?? _catalog;
+                if (expectedCatalog != _catalog)
+                {
+                    throw new InvalidOperationException($"Only one {nameof(ExportProvider)} can be created for a single test.");
+                }
 
-            var exportProvider = _currentExportProvider;
-            if (exportProvider != expected)
-            {
-                throw new InvalidOperationException($"Only one {nameof(ExportProvider)} can be created in the context of a single test.");
-            }
+                var expected = _expectedProviderForCatalog;
+                if (expected == null)
+                {
+                    expected = _exportProviderFactory.CreateExportProvider();
+                    expected = Interlocked.CompareExchange(ref _expectedProviderForCatalog, expected, null) ?? expected;
```

Comment:
Why the Interlocked* here? Is this trying to support a single test consuming multiple threads? It's clearly not multiple tests running in parallel (since this mechanism of using statics wouldn't work)...